exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -97,100 +104,105 @@
+        new ServiceRuntimeAdapter(serviceRuntime, viewFactory);
     boolean isAuditorNode = nodeType == EmulatedNodeType.AUDITOR;
-    UserServiceAdapter[] userServiceAdapters = serviceAdapters.values()
-        .toArray(new UserServiceAdapter[0]);
-    TimeProviderAdapter timeProviderAdapter =  timeProvider == null
-        ? null
-        : new TimeProviderAdapter(timeProvider);
-    long nativeHandle = nativeCreateTestKit(userServiceAdapters, isAuditorNode, validatorCount,
-        timeProviderAdapter);
-    return new TestKit(nativeHandle, serviceAdapters);
-  }
-
-  /**
-   * Returns a list of user service adapters created from given service modules.
-   */
-  private static Map<Short, UserServiceAdapter> toUserServiceAdapters(
-      List<Class<? extends ServiceModule>> serviceModules, Injector frameworkInjector) {
-    List<UserServiceAdapter> services = serviceModules.stream()
-        .map(s -> createUserServiceAdapter(s, frameworkInjector))
-        .collect(toList());
-    // UserServiceAdapters are removed, so this usage is a false signal for keeping getId()
-    return Maps.uniqueIndex(services, UserServiceAdapter::getId);
+    long nativeHandle = nativeCreateTestKit(serviceInstances, isAuditorNode, validatorCount,
+        timeServiceSpecs.toArray(new TimeServiceSpec[0]), serviceRuntimeAdapter);
+    return new TestKit(nativeHandle, timeServiceSpecs, serviceRuntime);
   }
 
-  /**
-   * Instantiates a service given its module and wraps in a UserServiceAdapter for the native code.
-   */
-  private static UserServiceAdapter createUserServiceAdapter(
-      Class<? extends ServiceModule> moduleClass, Injector frameworkInjector) {
-    try {
-      Supplier<ServiceModule> moduleSupplier = new ReflectiveModuleSupplier(moduleClass);
-      ServiceModule serviceModule = moduleSupplier.get();
-      Injector serviceInjector = frameworkInjector.createChildInjector(serviceModule);
-      return serviceInjector.getInstance(UserServiceAdapter.class);
-    } catch (IllegalAccessException e) {
-      throw new IllegalArgumentException("Cannot access the no-arg module constructor", e);
-    } catch (NoSuchMethodException e) {
-      throw new IllegalArgumentException("No no-arg constructor", e);
-    }
+  private static ServiceRuntime createServiceRuntime(Optional<Path> artifactsDirectory,
+                                              Optional<Integer> serverPort) {
+    String artifactsDir = artifactsDirectory.map(Path::toString).orElse("");
+    int port = serverPort.orElse(DEFAULT_SERVER_PORT);
+    return ServiceRuntimeBootstrap.createServiceRuntime(artifactsDir, port);
   }
 
-  private void populateServiceMap(Map<Short, UserServiceAdapter> serviceAdapters) {
-    serviceAdapters.forEach((id, serviceAdapter) -> services.put(id, serviceAdapter.getService()));
+  private static ViewFactory getViewFactory() {
+    Injector frameworkInjector = Guice.createInjector(new TestKitFrameworkModule());
```

Comment:
What is this injector for if it is **only** used to create a new ViewFactory, which is already provided by the other module, instantiated by the ServiceRuntimeBootstrap? If testkit creates an injector with the framework module, as suggest above, it will already have a binding for ViewFactory, hence it will be injected into SRA.